src/c++/perf_analyzer/client_backend/triton/triton_client_backend.cc

```diff
@@ -362,6 +367,153 @@ TritonClientBackend::ModelInferenceStatistics(
   return Error::Success;
 }
 
+Error
+TritonClientBackend::TritonMetrics(
+    triton::perfanalyzer::TritonMetrics& triton_metrics)
+{
+  try {
+    std::string metrics_endpoint_text{""};
+    AccessTritonMetricsEndpoint(metrics_endpoint_text);
+    ParseAndStoreTritonMetrics(metrics_endpoint_text, triton_metrics);
+  }
+  catch (const std::exception& e) {
+    return Error(e.what(), pa::GENERIC_ERROR);
+  }
+  return Error::Success;
+}
+
+void
+TritonClientBackend::AccessTritonMetricsEndpoint(
+    std::string& metrics_endpoint_text)
+{
+  CURL* curl{curl_easy_init()};
+  if (curl == nullptr) {
+    throw std::runtime_error("Error calling curl_easy_init()");
+  }
+
+  const auto write_function{
+      [](char* ptr, size_t size, size_t nmemb, std::string* userdata) {
+        userdata->append(ptr, size * nmemb);
+        return size * nmemb;
+      }};
+
+  curl_easy_setopt(curl, CURLOPT_URL, triton_metrics_url_.c_str());
+  curl_easy_setopt(
+      curl, CURLOPT_WRITEFUNCTION,
+      static_cast<size_t (*)(char*, size_t, size_t, std::string*)>(
+          write_function));
+  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &metrics_endpoint_text);
+
+  CURLcode res{curl_easy_perform(curl)};
+
+  if (res != CURLE_OK) {
+    throw std::runtime_error(
+        "curl_easy_perform() failed: " + std::string(curl_easy_strerror(res)));
+  }
+
+  long response_code{0};
+  curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
+
+  if (response_code != 200) {
+    throw std::runtime_error(
+        "Triton metrics endpoint curling did not succeed.");
+  }
+
+  curl_easy_cleanup(curl);
+}
+
+void
+TritonClientBackend::ParseAndStoreTritonMetrics(
+    const std::string& metrics_endpoint_text,
+    triton::perfanalyzer::TritonMetrics& triton_metrics)
+{
+  ParseAndStoreGPUUtilization(
+      metrics_endpoint_text, triton_metrics.gpu_utilization_per_gpu);
+  ParseAndStoreGPUPowerUsage(
+      metrics_endpoint_text, triton_metrics.gpu_power_usage_per_gpu);
+  ParseAndStoreGPUMemoryUsedBytes(
+      metrics_endpoint_text, triton_metrics.gpu_memory_used_bytes_per_gpu);
+}
+
+void
+TritonClientBackend::ParseAndStoreGPUUtilization(
+    const std::string& metrics_endpoint_text,
+    std::map<std::string, double>& gpu_utilization_per_gpu)
+{
+  std::regex gpu_utilization_regex(
+      R"(\nnv_gpu_utilization\{gpu_uuid\=\"([^"]+)\"\} (\d+\.?\d*))");
+  std::sregex_iterator gpu_utilization_begin{std::sregex_iterator(
+      metrics_endpoint_text.begin(), metrics_endpoint_text.end(),
+      gpu_utilization_regex)};
+
+  if (gpu_utilization_begin == std::sregex_iterator()) {
+    std::cerr << "WARNING: Triton 'nv_gpu_utilization' metric is missing or "
```

Comment:
If triton doesn't report this metric, this warning is going to emitted every time we query the endpoint which I think is bad UX.

Reply:
Change to only print warning on the first occurrence 👍